rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +134,228 @@ public void close(ISourceLocation loc) {
+                // If no new call to `calculate` has been made after `delay` has
+                // passed (i.e., `lastVersion` hasn't changed in the meantime),
+                // then run the calculation. Else, abandon this calculation.
+                if (latestVersion.get() == version) {
+                    return calculation.get();
+                } else {
+                    return CompletableFuture.completedFuture(Optional.<VersionedDiagnostics>empty());
+                }
+            }, delayed).thenCompose(Function.identity());
+        }
+
+        public void calculateAnalyzer(int version, Duration delay) {
+            calculateAnalyzerLatestResult = calculate(version, latestVersionCalculateAnalyzer, delay, () -> {
+                var analysis = summaryAnalyzer.calculateSummary();
+                var analysisTree = analysis.input;
+                var analysisMessages = analysis.output.get().thenApply(Supplier::get);
+                return analysisTree.thenCombine(analysisMessages, (tree, messages) -> {
+                    if (analysis.calculation.isInterrupted()) {
+                        return Optional.<VersionedDiagnostics>empty();
+                    } else {
+                        return Optional.of(reportDiagnostics(diagnosticsAnalyzer, tree, messages));
+                    }
+                });
+            });
+        }
+
+        /**
+         * The main challenge when running the builder is that it might produce
+         * a subset of the same diagnostics as the analyzer. Thus, to avoid
+         * showing the same diagnostics twice, a diff needs to be computed of
+         * the latest analyzer diagnostics and the current builder diagnostics.
+         * To do this in a reliable way, each builder calculation awaits the
+         * completion of the latest analyzer calculation, such that:
+         *   - the builder can use exactly the same AST as the analyzer;
+         *   - the builder has access to the diagnostics of the analyzer.
+         */
+        public void calculateBuilder(int version, Duration delay) {
+
+            // If an analyzer calculation hasn't been scheduled yet (ever), then
+            // it needs to be scheduled first.
+            if (calculateAnalyzerLatestResult == null) {
+                calculateAnalyzer(version, Duration.ZERO);
+                calculateBuilder(version, delay);
+            }
+
+            // If an analyzer calculation has been scheduled before, then the
+            // builder can await its completion.
+            else {
+                calculateAnalyzerLatestResult.thenAccept(result -> {
+
+                    // If the analyzer calculation wasn't abandoned (due to
+                    // debouncing) or interrupted, then diagnostics of the
+                    // analyzer are available, so proceed.
+                    if (result.isPresent()) {
+                        var analysis = result.get();
```

Comment:
shouldn't we check that the analysis we are basing ourselves on, is for the version we intent?

Reply:
In particular, we currently cannot guarantee that if an analyzer for version `n` begins, that it will actually use the parse tree of version `n` to do the analysis; when a concurrent `didChange` call occurs, it might use the parse tree of version `n+1`. Anyway, let's discuss this in a call if this route is preferrable.

Reply:
In particular, we currently cannot guarantee that if an analyzer for version `n` begins, that it will actually use the parse tree of version `n` to do the analysis; when a concurrent `didChange` call occurs, it might use the parse tree of version `n+1`. Anyway, let's discuss this in a call if this route is preferrable.